File: Try_GameEngine5/src/com/example/try_gameengine/action/InverseMovementInfoDecorator.java
package com.example.try_gameengine.action;

import java.util.List;

import android.util.Log;

public class InverseMovementInfoDecorator extends MovementDecorator{

	public InverseMovementInfoDecorator(MovementAction action) {
		this.action = action;
		this.copyMovementActionList = action.copyMovementActionList;
	}

	private MovementActionInfo coreCalculationMovementActionInfo(
			MovementActionInfo info) {
		info.setTotal(info.getTotal());
		info.setDelay(info.getDelay());
		info.setDx(-info.getDx());
		info.setDy(-info.getDy());
//		if(info.getRotationController()!=null)
//			info.getRotationController().setRotation(-info.getRotationController().getRotation());
		return info;
	}

	@Override
	public void start() {
		action.getAction().start();
	}

	@Override
	public MovementAction getAction() {
		return action.getAction();
	}

	@Override
	public String getDescription() {
		return "Double " + action.getDescription();
	}
	
	@Override
	public MovementAction initMovementAction(){	
		return initTimer();
	}

	@Override
	protected MovementAction initTimer(){ super.initTimer();

		if (this.getAction().getActions().size() == 0) {

				action.getAction().setInfo(getInfo());
				action.getAction().initTimer();

		} else {	
			this.getAction().initTimer();
			doIn();
		}
		return this;
	}

	@Override
	public MovementAction addMovementAction(MovementAction action) {
		getAction().addMovementAction(action);
		return this;
	}

	@Override
	protected void setActionsTheSameTimerOnTickListener() {
		getAction().setTimerOnTickListener(timerOnTickListener);
	}

	@Override
	public MovementActionInfo getInfo() {
		return coreCalculationMovementActionInfo(action.getInfo());
	}

	@Override
	public List<MovementAction> getCurrentActionList() {
		// TODO Auto-generated method stub
		return action.getCurrentActionList();
	}

	@Override
	public List<MovementActionInfo> getCurrentInfoList() {
		// TODO Auto-generated method stub
		return action.getCurrentInfoList();
	}
	
	@Override
	public List<MovementActionInfo> getMovementInfoList() {
		return action.getMovementInfoList();
	}
	
	@Override
	public void doIn(){
		action.doIn();
		this.getAction().getCurrentInfoList();
		int i = 0;
		for (MovementActionInfo info : this.getAction().currentInfoList) {
			Log.e("count", ++i + "");
			Log.e("info", info.getDx() + "");
			this.getAction().setInfo(info);
			coreCalculationMovementActionInfo(this.getAction().getInfo());
		}

//		for (MovementAction movementItem : this.getAction().movementItemList) {
//			movementItem.initTimer();
//		}
	}
}

File: Try_GameEngine5/src/com/example/try_gameengine/action/InverseMovementInfoDecorator.kt
package com.example.try_gameengine.action

class InverseMovementInfoDecorator(action: MovementAction) : MovementDecorator() {
    init {
        this.action = action
        //		this.copyMovementActionList = action.copyMovementActionList;
    }

    public override fun coreCalculationMovementActionInfo(
        action: MovementAction
    ): MovementAction {
        val info = action.getInfo()
        info.setTotal(info.getTotal())
        info.setDelay(info.getDelay())
        info.setDx(-info.getDx())
        info.setDy(-info.getDy())
        //		if(info.getRotationController()!=null)
//			info.getRotationController().setRotation(-info.getRotationController().getRotation());
        return action
    }

    override fun start() {
        action.getAction().start()
    }

    override fun getAction(): MovementAction {
        return action.getAction()
    }

    override fun getDescription(): String {
        return "Double " + action.getDescription()
    }

    public override fun initTimer(): MovementAction {
        super.initTimer()

        if (this.getAction()!!.getActions().size == 0) {
            action.getAction().setInfo(getInfo())
            action.getAction().initTimer()
        } else {
            this.getAction()!!.initTimer()
            //			doIn(null);
        }
        return this
    }

    override fun addMovementAction(action: MovementAction): MovementAction {
        getAction()!!.addMovementAction(action)
        return this
    }

    override fun setActionsTheSameTimerOnTickListener() {
        getAction()!!.setTimerOnTickListener(timerOnTickListener)
    }

    override fun getCurrentActionList(): MutableList<MovementAction> {
        // TODO Auto-generated method stub
        return action.getCurrentActionList()
    }

    override fun getCurrentInfoList(): MutableList<MovementActionInfo?> {
        // TODO Auto-generated method stub
        return action.getCurrentInfoList()
    }

    override fun getMovementInfoList(): MutableList<MovementActionInfo?> {
        return action.getMovementInfoList()
    } //	@Override
    //	protected List<MovementAction> doIn(MovementActionSet actionSet){
    //		List<MovementAction> actions = action.doIn(actionSet);
    //		this.getAction().getCurrentInfoList();
    //		int i = 0;
    //		for (MovementActionInfo info : this.getAction().currentInfoList) {
    //			Log.e("count", ++i + "");
    //			Log.e("info", info.getDx() + "");
    //			this.getAction().setInfo(info);
    //			coreCalculationMovementActionInfo(this.getAction());
    //		}
    //		
    //		return actions;
    //
    // /**/        for (MovementAction movementItem : this.getAction().movementItemList)
    // {
        // * /            movementItem.initTimer();
        // * /
    // } */
    //	}
}
